src/main/java/seedu/tasklist/model/ModelManager.java

```diff
@@ -73,29 +83,90 @@ public synchronized void deleteTask(ReadOnlyTask target) throws TaskNotFoundExce
         taskList.removeTask(target);
         updateFilteredListToShowIncomplete();
         indicateTaskListChanged();
+        if (callDidNotComeFromUndoCommand())
+            addToUndoStack(UndoCommand.DEL_CMD_ID, (Task) target);
     }
-
+    @Override
+    public void deleteTaskUndo(ReadOnlyTask target) throws TaskNotFoundException {
+        taskList.removeTask(target);
+        updateFilteredListToShowIncomplete();
+        indicateTaskListChanged();
+    }
+    
     @Override
     public synchronized void addTask(Task task) throws UniqueTaskList.DuplicateTaskException {
         taskList.addTask(task);
         updateFilteredListToShowIncomplete();
         indicateTaskListChanged();
+        if (callDidNotComeFromUndoCommand())
+            addToUndoStack(UndoCommand.ADD_CMD_ID, task);
+    }
+
+    @Override
+    public void addTaskUndo(Task task) throws UniqueTaskList.DuplicateTaskException {
+        taskList.addTask(task);
+        updateFilteredListToShowIncomplete();
+        indicateTaskListChanged();
+    }
+    
+    @Override
+    public synchronized void updateTask(Task taskToUpdate, TaskDetails taskDetails, StartTime startTime,
+            EndTime endTime, Priority priority, UniqueTagList tags) throws UniqueTaskList.DuplicateTaskException {
+        Task originalTask = new Task(taskToUpdate);
+        taskList.updateTask(taskToUpdate, taskDetails, startTime, endTime, priority);
+        updateFilteredListToShowIncomplete();
+        indicateTaskListChanged();
+        if (callDidNotComeFromUndoCommand()) {
```

Comment:
is this check currently necessary? it seems the if statement always evaluates to true